src/strings/gb18030.c

```diff
@@ -0,0 +1,346 @@
+64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
+64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
+64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64};
+
+const MVMint32 gb18030_two_byte_upper_bound[126] = {
+254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,
+254,254,254,254,254,254,254,254,254,254,254,254,254,252,254,243,246,245,241,233,
+239,160,160,160,160,160,160,254,254,254,254,254,254,254,254,254,254,254,254,254,
+254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,
+254,254,254,254,254,254,249,254,254,254,254,254,254,254,254,254,254,254,254,254,
+254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,160,
+160,160,160,160,160,159};
+
+MVMint32 gb18030_valid_check_len2(MVMint32 c_1, MVMint32 c_2) {
+    /* This function serves like a 'first stage check' of c_1 and c_2.
+       It eliminates most of the invalid combinations of c_1 and c_2, 
+       but for code simplicity and to avoid lots of if-else here,
+       ther remaining invalid combinations will be processed 
+       in the conversion table. */
+    if (c_1 < 0x81 || c_1 > 0xfe) return 0;
+    c_1 -= 0x81;
+    return gb18030_two_byte_lower_bound[c_1] <= c_2 && c_2 <= gb18030_two_byte_upper_bound[c_1] ? 1 : 0;
+}
+
+MVMint32 gb18030_valid_check_len4(MVMint32 c_1, MVMint32 c_2, MVMint32 c_3, MVMint32 c_4) {
+    if ((0x81 <= c_1 && c_1 <= 0x83) || (c_1 == 0x84 && c_2 == 0x30)) {
+        return (0x30 <= c_2 && c_2 <= 0x39) && (0x81 <= c_3 && c_3 <= 0xfe) && (0x30 <= c_4 && c_4 <= 0x39) ? 1 : 0;
+    } else if (c_1 == 0x84 && c_2 == 0x31) {
+        return (0x81 <= c_3 && c_3 <= 0xa4) && (0x30 <= c_4 && c_4 <= 0x39) ? 1 : 0;
+    }
+    return 0;
+}
+
+MVMint32 gb18030_valid_check_len4_first2(MVMint32 c_1, MVMint32 c_2) {
+    return (((0x81 <= c_1 && c_1 <= 0x83) || (c_1 == 0x84 && c_2 == 0x30)) && (0x30 <= c_2 && c_2 <= 0x39))	|| (c_1 == 0x84 && c_2 == 0x31) ? 1 : 0;
+}
+
+MVMString * MVM_string_gb18030_decode(MVMThreadContext *tc, const MVMObject *result_type, const char *gb18030, size_t bytes) {
+    size_t i, result_graphs;
+
+    MVMString *result = (MVMString *)REPR(result_type)->allocate(tc, STABLE(result_type));
+
+    result->body.storage_type = MVM_STRING_GRAPHEME_32;
+    result->body.storage.blob_32 = MVM_malloc(sizeof(MVMGrapheme32) * bytes);
+
+    result_graphs = 0;
+
+    for (i = 0; i < bytes; i++) {
+        if (0 <= gb18030[i] && gb18030[i] <= 127) {
+            if (gb18030[i] == '\r' && i + 1 < bytes && gb18030[i + 1] == '\n') {
+                result->body.storage.blob_32[result_graphs++] = MVM_nfg_crlf_grapheme(tc);
+                i++;
+            }
+            else {
+                result->body.storage.blob_32[result_graphs++] = gb18030[i];
+            }
+        }
+        else {
+            if (i + 1 < bytes) {
+            //  GB18030 codepoint of length 2
```

Comment:
We prefer using `/* comment */` style comments in MoarVM so if you can change all the `//` comments to use the `/* */` style that would be great.